### ISSUE TITLE
model.fit_generator is too slow when dataset is large

### ISSUE THREAD

**Issue by granthst** [NTMC-Community/MatchZoo#40]
//training is too slow when the dataset is large 
genfun = generator.get_batch_generator()
history = model.fit_generator(
                    genfun,
                    steps_per_epoch = display_interval,
                    epochs = 1,
                    shuffle=False,
                    verbose = 1
                )
<img width="639" alt="dingtalk20180109160210" src="https://user-images.githubusercontent.com/6413758/34710478-83ac7b2e-f556-11e7-879a-ae4f5ad077c7.png">

**Comment by granthst** [comment]
program virtually freezes during either of there two steps, is there anything wrong with the pair_generator.py file implementation?

**Comment by granthst** [comment]
I am using cpu for training and my laptop has 16gRAM, during training, the python program occupies about all of my RAM

**Comment by granthst** [comment]
Has anyone try this on a relatively large dataset, bigger than the wikiqa data set?

**Comment by uduse** [comment]
You're using your CPU to train?

**Comment by akionakamura** [comment]
I'm wondering if this actually has to do with a wrongly set `steps_per_epoch` argument in the `fit_generator` method. Note how the code is using `display_interval` value as the `steps_per_epoch` argument ([code](https://github.com/faneshion/MatchZoo/blob/55b986aff1cafe03c5fc28f1db5dad22a725dbd8/matchzoo/main.py#L148)), even though the documentation says [display the train loss with test performance each "display_interval" iters](https://github.com/faneshion/MatchZoo/blob/55b986aff1cafe03c5fc28f1db5dad22a725dbd8/matchzoo/models/README#L13).

If you set `display_interval = 1`, you're also setting that each epoch only contains a single batch from the generator, making a epoch very quick. Therefore, the code spends most of the time in boilerplates, instead of actual training.

According to Keras [documentation](https://keras.io/models/model/#fit_generator):

> steps_per_epoch: Total number of steps (batches of samples) to yield from generator before declaring one epoch finished and starting the next epoch. It should typically be equal to the number of samples of your dataset divided by the batch size. Optional for Sequence: if unspecified, will use the len(generator) as a number of steps.

So if your `display_interval` value is too small, you might actually be consuming very little of you (large) dataset per epoch. Ideally, the number of step should cover the entire dataset, according to the size of the batch as well. The `num_iters` sets the number of epochs.

Furthermore, the current code displays the train loss at every iterations, so you can directly set the `display_interval` as your `steps_per_epoch` without losing the display-control functionality.

**Comment by yangliuy** [comment]
@akionakamura I think most of your descriptions are correct. If you take a look at the https://github.com/faneshion/MatchZoo/blob/master/matchzoo/inputs/pair_generator.py , you can find the implementation of batch generation process. There are two modes : use_iter=True and use_iter=False. The process of batch generation under these two modes are different. Even you set use_iter=False, the iteration of all pairs in a batch is also stochastic if you noticed the line 125 of this file as following: 

d1, d2p, d2n = random.choice(self.pair_list)

So let's say we set a small display_interval=10, batch_size=200, num_iters=500, then for each iteration you will iter 10*200=2000 randomly sampled pairs. After 500 iterations, you will iter 10*200*500=10^6 randomly sampled pairs. So if your training data is large and you set use_iter=False&display_interval=10, you can choose a larger num_iters to sample more training pairs. So you can see the concept of ''iterations'' in MatchZoo is different from ''epoches'' in some papers. One epoch means an iteration of all the training data one time in those papers. But this is not the case for the implementation in MatchZoo.

If you set use_iter=True, then there is a different story. You can find this by checking the code of https://github.com/faneshion/MatchZoo/blob/master/matchzoo/inputs/pair_generator.py  .

**Comment by bwanglzu** [comment]
This will be an enhancement of 2.0

**Comment by kinglai** [comment]
@yangliuy  
thus that mean, you can just use a little data each iter ? But how can you make sure the model's convergence?  

And when set use_iter=True, the training time per iter is much longer when the training data is much larger, but the training data per iter is the same , it seem's a little ridiculous, cause the time is waste in sample data ?
And the use the param num_iters is depand on the size of training data ,but not the model ? It seems not so reasonable.

**Comment by bwanglzu** [comment]
@faneshion @yangliuy please take a look at @kinglai 's question.
